lib/state/state.js

```diff
@@ -292,76 +305,76 @@ State.prototype.findActors = function (actorIds, cb) {
 };
 
 
-State.prototype.emitEvents = function (cb) {
-	if (!mage.core.msgServer) {
-		logger.warning('Cannot emit events without msgServer set up.');
-		return cb();
-	}
-
-	var events;
+function parseEvents(isError, events) {
+	return events
+		.filter((entry) => !isError || entry.alwaysEmit)
+		.map((entry) => entry.evt);
+}
 
-	// this function emits all events to players who are not state.actorId
 
-	events = this.broadcastEvents;
+State.prototype.emitEvents = function (isErrorState, cb) {
+	// Reset the state's events store
+	const events = this.otherEvents;
+	this.otherEvents = {};
 
-	if (events) {
-		this.broadcastEvents = undefined;
-		mage.core.msgServer.broadcast('[' + events.join(',') + ']');
-	}
+	const broadcastEvents = this.broadcastEvents;
+	this.broadcastEvents = [];
 
-	events = this.otherEvents;
+	// Extract the list of actorIds we will be sending messages
+	// to directly (not by broadcast)
+	const actorIds = Object.keys(events);
 
-	if (!events) {
+	// If we have nothing to do at all, return immediately
+	if (actorIds.length === 0 && broadcastEvents.length === 0) {
 		return cb();
 	}
 
-	this.otherEvents = undefined;
+	// We either have messages to broadcast, or messages to send
+	// to a number of actorIds; confirm that msgServer is present,
+	// or else return an error
+	if (!mage.core.msgServer) {
+		return cb(new StateError('Cannot emit events without msgServer set up.', {
+			events: events,
+			broadcastEvents: broadcastEvents
+		}));
+	}
+
+	// Send all broadcast events
+	const parsedBroadcastEvents = parseEvents(isErrorState, broadcastEvents);
 
-	var actorIds = Object.keys(events);
+	if (parsedBroadcastEvents.length > 0) {
+		mage.core.msgServer.broadcast('[' + parsedBroadcastEvents.join(',') + ']');
+	}
 
+	// If we have no other messages to send,
+	// return early; this is to avoid running
+	// additional code that would only delay the return
 	if (actorIds.length === 0) {
 		return cb();
 	}
 
-	this.lookupAddresses(actorIds, function (error, addresses) {
+	lookupAddresses(this, actorIds, function (error, addresses) {
 		if (error) {
 			return cb(error);
 		}
 
 		for (var i = 0; i < actorIds.length; i += 1) {
 			var actorId = actorIds[i];
-			var entries = events[actorId];
+			var entries = parseEvents(isErrorState, events[actorId]);
```

Comment:
Since `parseEvents` has a relevant cost, it might be good to `continue` early if there is no `address` for this `actorId`, which will be *very* common. We very often emit events to a user's "friends" etc, without knowing if that friend is online. Assuming most of the day users are *not* playing, we will have more misses than hits.

Reply:
Fixed